fitbenchmarking/parsing/tests/test_parsers.py

```diff
@@ -0,0 +1,235 @@
+    expectation.data_y = np.array(expectation_dict['data_y'])
+    expectation.data_e = expectation_dict['data_e']
+    if expectation.data_e is not None:
+        expectation.data_e = np.array(expectation.data_e)
+    expectation.functions = expectation_dict['functions']
+    expectation.starting_values = expectation_dict['starting_values']
+    expectation.value_ranges = expectation_dict['value_ranges']
+
+    return expectation
+
+
+class TestParsers:
+    """
+    A class to hold the tests for parametrized testing of parsers.
+    """
+
+    params = generate_test_cases()
+
+    def test_parsers(self, file_format, test_file, expected):
+        """
+        Test that the parser correctly parses the specified input file.
+
+        :param file_format: The name of the file format
+        :type file_format: string
+        :param test_file: The path to the test file
+        :type test_file: string
+        """
+        assert (test_file is not None), \
+            'No test file for {}'.format(file_format)
+
+        # Test import
+        module = import_module(name='.{}_parser'.format(file_format),
+                               package='fitbenchmarking.parsing')
+
+        parser = getmembers(module, lambda m: (isclass(m)
+                                               and not isabstract(m)
+                                               and issubclass(m, Parser)
+                                               and m is not Parser))[0][1]
+
+        # Test parse
+        with parser(test_file) as p:
+            fitting_problem = p.parse()
+
+        # Allow for problems not supporting certain test cases
+        # (e.g. value_ranges)
+        if (fitting_problem.name == 'NA'):
+            return
+
+        # Check against expected
+        fitting_problem.verify()
+
+        # functions and equation need to be done seperately as they can't be
+        # generic across problem types.
+        # similarly starting_values uses the param name so must be checked
+        # separately
+        for attr in ['name', 'data_x', 'data_y', 'data_e', 'start_x', 'end_x',
+                     'value_ranges']:
+            parsed_attr = getattr(fitting_problem, attr)
+            expected_attr = getattr(expected, attr)
+            equal = (parsed_attr == expected_attr)
+            if isinstance(equal, np.ndarray):
+                equal = equal.all()
+            assert (equal), '{} was parsed incorrectly.'.format(attr) \
+                + '{} != {}'.format(parsed_attr, expected_attr)
+
+        # Check starting_values, but ignore the param names
+        for a, e in zip(fitting_problem.starting_values,
+                        expected.starting_values):
+            assert (a[1] == e[1]), 'starting_values were parsed incorrectly.'
+
+        # Check functions length but nothing else here.
+        assert (len(fitting_problem.functions) == len(expected.functions)), \
+            'functions were parsed incorrectly.'
```

Comment:
The fact that you have three asserts relating to different things (even though they share a setup) probably means you should have three different tests.  I wouldn't worry about changing it in this instance, I only mention it as a general rule.

Reply:
Ok, I'll keep it in mind.
On that note though, the first assert is in a loop. Would you advise those to all be split out as well?

Reply:
I was debating that and probably strictly speaking they should, as I think they are all governed by different methods within the parser.

I think each unit test should test a specific bit of functionality.  At the moment, I think this test fails if one of the parser methods (which combine to form `parse`) is invalid, but also if all of them are.  So you are getting less information than if you split them out.

Admittedly there will always be a little bit greyness as to what should go into a single unit test.  I think having multiple asserts is fine, as long as they are just providing ways of testing the same functionality.

Reply:
Ok, I'll leave it for now and try to find better ways of doing this kind of thing in future